### PR TITLE
fix(backup-chain): next log backup can have the same `LastLsn`

### PR DESCRIPTION
**Pull request by AdityaNPL**
(It looks like it made loads of white space changes automatically - pls ignore this while reviewing.)

I came across an instance where the `LastLsn` was the same for 2 backups. Hence, this check was filtering the log backup out.

image

### CODE REVIEW

File: src/BackupChain.cs [hunk 1]
```diff
@@ -54,51 +54,51 @@ private BackupChain(IList<BackupMetadata> recentBackups)
...
+      // get all the stripes of this backup
+      return diffBackupsOrdered.Where(diffBackup => diffBackup.LastLsn == targetLastLsn);
+    }
+
+    private static IEnumerable<BackupMetadata> NextLogBackup(IEnumerable<BackupMetadata> backups,
+      BackupMetadata prevBackup)
+    {
+      // also gets all the stripes of the next backup
+      return backups.Where(b => b.BackupType == BackupFileTools.BackupType.Log &&
+                                prevBackup.LastLsn >= b.FirstLsn && prevBackup.LastLsn <= b.LastLsn);
```
Comment (AdityaNPL) [inline src/BackupChain.cs:95]: I'm wondering, if this is a simpler check?
```csharp
prevBackup.FirstLsn <= b.FirstLsn && prevBackup.LastLsn <= b.LastLsn
```

Comment (alexirion10) [reply src/BackupChain.cs:95]: no, the point of chaining is to check first_lsn vs last_lsn

**Review by alexirion10 — COMMENTED** [review]
Not sure what scenario caused this, but ok.  Take a look at the failing unit test plz

**Comment by AdityaNPL** [comment]
Unit test:
> System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown.

huh, that's weird - checking

**Comment by AdityaNPL** [comment]
uggh, infinite loop while building the chain cause it's matching the same log backup over and over again

File: src/BackupMetadata.cs [hunk 2]
```diff
@@ -5,31 +5,45 @@ namespace AgDatabaseMove
   using SmoFacade;
 
 
-  /// <summary>
-  ///   Occasionally we wind up with the same entry for a backup on multiple instance's msdb.
-  ///   For now we'll consider these backups to be equal despite their file location,
-  ///   but perhaps there's value in being able to look for the file in multiple locations.
-  /// </summary>
-  public class BackupMetadataEqualityComparer : IEqualityComparer<BackupMetadata>
+  public class BackupEqualityComparer : IEqualityComparer<BackupMetadata>
```
Comment (AdityaNPL) [inline src/BackupMetadata.cs:8]: Needed to borrow this from the other PR to fix the issue here:
https://github.com/factset/AgDatabaseMove/pull/101/files#diff-0b43e3447245b624777b212abcd1a5a74ed28b239eb4797e66e7ce06b3288e62R8

**Comment by AdityaNPL** [comment]
Fixed the infinite loop by making sure it isn't getting the same backup as the last one

---

Tested on a MSSQL server with the lastLSNs equal for two of the backups
Also the unit tests should run fine now

**Review by alexirion10 — APPROVED** [review]

**Pull request merged**